hyperspy/axes.py

```diff
@@ -559,22 +559,33 @@ def value2index(self, value, rounding=round):
         ------
         ValueError
             If any value is out of the axis limits.
-
+            If value is NaN
+            If value is an array containing at least 1 NaN value
+            If value is an array containing at least 1 out-of-bound value
         """
         if value is None:
             return None
-        if self.low_value <= value <= self.high_value:
-            return (np.abs(self.axis - value)).argmin()
+        #Should evaluate on both arrays and scalars. Raises error if there are
+        #nan values in array
+        if np.all((value >= self.low_value)*(value <= self.high_value)):
+            #initialise the index same dimension as input, force type to int
+            index = np.asarray(value).astype(int).copy()
+            #assign on flat, iterate on flat.
+            for i,v in enumerate(np.asarray(value).flat):
+                index.flat[i] = (np.abs(self.axis - v)).argmin()
+            #Squeezing to get a scalar out if scalar in. See squeeze doc
+            return np.squeeze(index)[()]
         else:
-            index = int(value)
-            if self.size > index >= 0:
-                return index
-            else:
-                raise ValueError(
-                    f'The value {value} is out of the limits '
-                    f'[{self.low_value:.3g}-{self.high_value:.3g}] of the '
-                    f'"{self._get_name()}" axis.'
-                    )
+            ### I commented out this because it seemed dangerous
+            # index = int(value)
```

Comment:
Yes, I agree and I ignore it because, I think that it may be better to discuss this in a separate PR. If you want to fix this in this PR, the `value2index` of `UniformDataAxis` needs to be changed consistently.

Reply:
This change already brings consistency with UniformDataAxis.

As implemented, FunctionalDataAxis.value2index allows for something very wrong to happen, ie it interpretates the value as an index if it can interpretate it as an int.

_Voyez plutot_

```python
>>> from hyperspy.axes import FunctionalDataAxis,UniformDataAxis
>>> uax = UniformDataAxis(size=101.0,offset=0.0,scale=0.01)
>>> uax.value2index(50)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-5cf2ab8da91f> in <module>
      1 uax = UniformDataAxis(size=101.0,offset=0.0,scale=0.01)
----> 2 uax.value2index(50)

c:\users\nicolastappy\documents\git\hyperspy\hyperspy\axes.py in value2index(self, value, rounding)
   1177                 return index
   1178             else:
-> 1179                 raise ValueError("The value is out of the axis limits")
   1180 
   1181     def update_axis(self):

ValueError: The value is out of the axis limits
```

Which is expected. Now
```python
>>> nuax = FunctionalDataAxis(expression="x**2",x=uax)
>>> nuax.value2index(69.69)
69
>>> nuax.value2index(99)
99
```
Which is absolutely wrong since `nuax` spans 0 to 1 like `uax`. This works with `uax` because double values are converted to indices before. I think this was placed here with the intent to let value2index fall back on `int==index`
like `isig[]` slices but in fact, it's simply wrong.

I am fine making another PR about this, but I'd rather have this fixed ASAP.

Reply:
I looked at it in more depth, and indeed this is no reason for this to be there. When looking at it quickly, I used the `UniformDataAxis.value2index` as a reference, but it is different, because the index is calculated before. I was expecting this to follow a logic but I looks to me that when this was implemented, this fell through the cracks!

I would suggest to remove it because the if block before that cover scalar and arrays. Would it be possible to remove `UniformDataAxis.value2index` and use `BaseDataAxis.value2index` instead?

Reply:
I will do that then.

I will then also implement the "rounding" argument (which I actually forgot in the first place)

Reply:
> I would suggest to remove it because the if block before that cover scalar and arrays. Would it be possible to remove `UniformDataAxis.value2index` and use `BaseDataAxis.value2index` instead?

This failed pretty badly.  `UniformDataAxis `inherits from `UnitConversion ` and `UniformDataAxis.value2index` uses some of `UnitConversion` methods to parse string inputs.

Reply:
Agree, let's finish this PR and #2747 and then it would be easier to remove `UniformDataAxis.value2index`.

Reply:
Add to todo list in #2398?

Reply:
On a second thought, I think that keeping the possibility of overriding BaseDataAxis value2index is still a good option. Thinking about axes in a unified manner is a good option, because it allows users to easily adapt hs to their needs. However, a few specific cases (UniformDataAxis being the most common) will cover most of the experimental needs. For those, keeping override of BaseDataAxis as a standard encourages code optimisation

Here is a picture of UniformDataAxis.value2index() vs BasedataAxis.value2index(), which, I think, is self-explanatory
 
![image](https://user-images.githubusercontent.com/37581871/119500999-b6706600-bd68-11eb-8ddb-7c07e0946fab.png)

Reply:
Do you know which part is slow? If so, can it be fixed?

Reply:
It cannot unfortunately. np.argmin(x-v) is the nominal strategy for finding a closest value using numpy, as far as I know.
Still it is O(N). In comparison, explicit inversion of the axis function as in UniformDataAxis is probably O(1) or something close. 

I don't see any way to improve the speed of this. This needs to be applicable in DataAxis, which is purely value-defined.

A solution can be found in FunctionalDataAxis though, provided the function can be inverted.

Reply:
I agree to try.

I will try to have numba'ed functions for rounding and checking if values are within limits. numba_closest_index_round, numba_closest_index_floor, numba_closest_index_ceil, numba_check_limits and let the BaseDataAxis.value2index take care of input formatting

Reply:
Ok it kinda works with numba.

However the argument stays the same. I could make it ~10 times faster with numba, I can probably make `UniformDataAxis.value2index() `10times faster as well

![image](https://user-images.githubusercontent.com/37581871/119814835-273f8b80-beeb-11eb-9b08-3888a354abf5.png)

Reply:
It is definitely nice to have this optimized, but I agree to your above statement that it does not hurt to override certain methods if they can be implemented more efficiently.

Reply:
@ecripre I would propose to finish this PR and we can further optimize `value2index` in later PRs, what do you think?

Reply:
> @ecripre I would propose to finish this PR and we can further optimize `value2index` in later PRs, what do you think?

Yes!